File: WoofIOS/SwiftPackages/Domain/UseCase/Sources/UseCase/Dogs/GetDogsFeedUseCase.swift
import Combine
import DI
import Models
import Repositories

@Factory
public final class GetDogsFeedUseCase {
    private let dogsRepository: DogsRepository
    private let moderationRepository: ModerationRepository

    public func callAsFunction() -> AnyPublisher<DogsFeed, Never> {
        Publishers.CombineLatest(
            dogsRepository.dogsFeed,
            moderationRepository.blockedDogIds
        )
        .map { feed, blockedIds in
            DogsFeed(
                dogs: feed.dogs.filter { !blockedIds.contains($0.id) },
                total: feed.total - blockedIds.count
            )
        }
        .removeDuplicates()
        .eraseToAnyPublisher()
    }
}

File: WoofAndroid/domain/usecase/src/main/kotlin/com/perrystreet/woof/usecase/dogs/GetDogsFeedUseCase.kt
package com.perrystreet.woof.usecase.dogs

import com.perrystreet.woof.models.dog.DogsFeed
import com.perrystreet.woof.repositories.dogs.DogsRepository
import com.perrystreet.woof.repositories.moderation.ModerationRepository
import io.reactivex.rxjava3.core.Observable
import org.koin.core.annotation.Factory

@Factory
class GetDogsFeedUseCase(
    private val dogsRepository: DogsRepository,
    private val moderationRepository: ModerationRepository,
) {
    operator fun invoke(): Observable<DogsFeed> =
        Observable.combineLatest(
            dogsRepository.dogsFeed,
            moderationRepository.blockedDogIds,
        ) { feed, blockedIds ->
            DogsFeed(
                dogs = feed.dogs.filterNot { it.id in blockedIds },
                total = feed.total - blockedIds.size,
            )
        }.distinctUntilChanged()
}
